File: models/device.go
package models

import (
	"gorm.io/gorm"
)

//DeviceModel 设备模型
type DeviceModel struct {
	gorm.Model
	Name            string            `gorm:"size:50"`   //设备名称
	Desc            string            `gorm:"size:5000"` //设备描述
	DeviceType      uint8             //设备类型，0:默认设备,1:摄像头
	CameraCompany   uint8             //摄像头厂家，0：海康威视，1：大华
	AttributeModels []*AttributeModel //属性模型
	TelemetryModels []*TelemetryModel //遥测模型
}

//AttributeModel 属性模型
type AttributeModel struct {
	gorm.Model
	DeviceModelID uint   `gorm:"unique_index:attribute_model_uix"`         //设备模型id
	Key           string `gorm:"unique_index:attribute_model_uix;size:50"` //设备模型下唯一
	Name          string `gorm:"size:50"`                                  //名称
}

//TelemetryModel 遥测模型
type TelemetryModel struct {
	gorm.Model
	DeviceModelID uint    `gorm:"unique_index:telemetry_model_uix"`         //设备模型id
	Key           string  `gorm:"unique_index:telemetry_model_uix;size:50"` //设备模型下唯一
	Name          string  `gorm:"size:50"`                                  //名称
	Factor        float64 //系数
	Unit          string  `gorm:"size:10"` //单位
	UnitName      string  `gorm:"size:10"` //单位名称
	Scale         uint8   //小数位数
}

//Device 设备
type Device struct {
	gorm.Model
	DeviceModelID uint //设备模型id
	DeviceModel   DeviceModel
	Name          string       `gorm:"size:50;NOT NULL;"` //设备名称
	IP            string       `gorm:"size:50;NOT NULL;"` //设备ip
	AccessKey     string       `gorm:"size:12;NOT NULL;"` //设备key hashid(id,16)
	SecretKey     string       `gorm:"size:32;NOT NULL;"` //设备访问密钥,randString()
	Port          uint16       //设备端口
	Username      string       `gorm:"size:32;"`
	Password      string       `gorm:"size:32;"`
	LivePath      string       `gorm:"size:100"` //直播路径
	Attributes    []*Attribute //属性，由采集程序启动时注册
	Telemetries   []*Telemetry //遥测，由采集程序按一定频率上传
}

//Attribute 属性
type Attribute struct {
	gorm.Model
	DeviceID         uint `gorm:"unique_index:attribute_uix"` //设备模型id
	AttributeModelID uint `gorm:"unique_index:attribute_uix"`
	AttributeModel   AttributeModel
	Value            string `gorm:"size:50;NOT NULL;"`
}

//Telemetry 遥测
type Telemetry struct {
	gorm.Model
	DeviceID         uint `gorm:"unique_index:telemetry_uix"` //设备id
	Device           Device
	TelemetryModelID uint `gorm:"unique_index:telemetry_uix"`
	TelemetryModel   TelemetryModel
}

//DeviceDashboard 设备仪表盘
type DeviceDashboard struct {
	gorm.Model
	Name        string `gorm:"size:50"`
	IsVisible   bool
	DeviceType  uint8 //设备类型，0:默认设备,1:摄像头
	Telemetries []*DeviceDashboardTelemetry
	Cameras     []*DeviceDashboardCamera
}

//DeviceDashboardTelemetry 仪表盘遥测
type DeviceDashboardTelemetry struct {
	gorm.Model
	DeviceDashboardID uint
	TelemetryID       uint
	Telemetry         Telemetry
}

//DeviceDashboardCamera 仪表盘摄像头
type DeviceDashboardCamera struct {
	gorm.Model
	DeviceDashboardID uint
	DeviceID          uint
	Device            Device
}

File: web/src/models/device.ts

export interface IDeviceModel {
    id: number
    name: string
    desc: string
    deviceType: number
    cameraCompany: number
    createdAt: number
    updatedAt: number
}

export interface IDevice {
    id: number
    name: string
    ip: string
    port: number
    accessKey: string
    secretKey: string
    deviceModelID: number
    deviceModel: IDeviceModel
    health?: number
    username: string
    password: string
    createdAt: number
    updatedAt: number
}

export interface IDeviceDashboard {
    id: number
    name: string
    isVisible: boolean
    telemetries: IDeviceDashboardTelemetry[]
    cameras: IDeviceDashboardCamera[]
    deviceType: number
    createdAt: number
    updatedAt: number
}

export interface IDeviceDashboardTelemetry {
    id: number
    deviceID: number
    deviceName: string
    name: string
    value?: number
    factor: number
    unit: string
    scale: number
    telemetryID: number
}

export interface IDeviceDashboardCamera {
    id: number
    deviceID: number
    deviceName: string
}

export interface INewDeviceModel {
    name: string
    deviceType: number
    desc: string
    cameraCompany: number
}

export interface IUpdateDeviceModel {
    id: number
    name: string
    desc: string
}

export interface INewDevice {
    name: string
    deviceModelID: number
    ip: string
    port: number
    username: string
    password: string
}

export interface IUpdateDevice {
    id: number
    name: string
    ip: string
    port: number
    username: string
    password: string
}

export interface INewDeviceDashboard {
    name: string
    isVisible: boolean
    deviceType: number
}

export interface IUpdateDeviceDashboard {
    id: number
    name: string
    isVisible: boolean
}

export interface INewDeviceDashboardTelemetry {
    deviceDashboardID: number
    telemetryIDs: number[]
}

export interface INewDeviceDashboardCamera {
    deviceDashboardID: number
    deviceIDs: number[]
}

export interface INewAttributeModel {
    key: string
    name: string
}

export interface IUpdateAttributeModel {
    id: number
    key: string
    name: string
}

export interface INewTelemetryModel {
    key: string
    name: string
    factor: number
    unit: string
    unitName: string
    scale: number
}

export interface IUpdateTelemetryModel {
    id: number
    key: string
    name: string
    factor: number
    unit: string
    unitName: string
    scale: number
}
